x/amm/keeper/source.go

```diff
@@ -157,46 +138,48 @@ func (k Keeper) AfterPoolOrdersExecuted(ctx sdk.Context, pool types.Pool, result
 		currentSqrtPrice := utils.DecApproxSqrt(poolState.CurrentPrice)
 		var nextSqrtPrice, nextPrice sdk.Dec
 		max = false
-		if result.Order.ExecutableQuantity(result.Order.Price).IsZero() { // Fully executed
-			nextSqrtPrice = utils.DecApproxSqrt(result.Order.Price)
-			nextPrice = result.Order.Price
+		if i < len(results)-1 || result.ExecutableQuantity().LTE(utils.SmallestDec) {
+			nextSqrtPrice = utils.DecApproxSqrt(result.Price())
+			nextPrice = result.Price()
 			max = true
 		} else { // Partially executed
 			// TODO: fix nextSqrtPrice?
 			nextSqrtPrice = types.NextSqrtPriceFromOutput(
-				currentSqrtPrice, poolState.CurrentLiquidity, result.Paid.Amount, result.Order.IsBuy)
+				currentSqrtPrice, poolState.CurrentLiquidity, result.PaidWithoutFee(), isBuy)
 			nextPrice = nextSqrtPrice.Power(2)
 		}
 
-		var expectedAmtIn sdk.Int
-		if result.Order.IsBuy {
-			expectedAmtIn = types.Amount0Delta(
-				currentSqrtPrice, nextSqrtPrice, poolState.CurrentLiquidity)
+		var expectedAmtIn sdk.Dec
+		if isBuy {
+			expectedAmtIn = types.Amount0DeltaRoundingDec(
+				currentSqrtPrice, nextSqrtPrice, poolState.CurrentLiquidity, true)
 		} else {
-			expectedAmtIn = types.Amount1Delta(
+			expectedAmtIn = types.Amount1DeltaDec(
 				currentSqrtPrice, nextSqrtPrice, poolState.CurrentLiquidity)
 		}
-		denomIn := pool.DenomIn(isBuy)
-		amtInDiff := result.Received.AmountOf(denomIn).Sub(expectedAmtIn)
+		amtInDiff := result.Received().Sub(expectedAmtIn)
 		if amtInDiff.IsPositive() {
-			fee := sdk.NewCoin(denomIn, amtInDiff)
+			denomIn := pool.DenomIn(isBuy)
+			fee, _ := sdk.NewDecCoinFromDec(denomIn, amtInDiff).TruncateDecimal()
 			accruedRewards = accruedRewards.Add(fee)
-			feeGrowth := sdk.NewDecCoinFromDec(fee.Denom, fee.Amount.ToDec().
-				MulTruncate(types.DecMulFactor).
-				QuoTruncate(poolState.CurrentLiquidity.ToDec()))
+			feeGrowth := sdk.NewDecCoinFromDec(
+				fee.Denom, fee.Amount.ToDec().
+					MulTruncate(types.DecMulFactor).
+					QuoTruncate(poolState.CurrentLiquidity.ToDec()))
 			poolState.FeeGrowthGlobal = poolState.FeeGrowthGlobal.Add(feeGrowth)
 		} else if amtInDiff.IsNegative() { // sanity check
-			//panic(amtInDiff)
+			panic(amtInDiff)
```

Comment:
I think this part is likely to happen, I will share the detailed scenario again as soon as it is sorted out